README.md
Copyright (2003-2010) Mediterranean, Bro3

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 59 Temple
Place, Boston, MA 02111.
</pre>

MeD's Movie Manager is a simple to use, yet customizable, movie manager. It gets
information about movies and series from IMDb.

<div class="section" id="system-requirements">

# System Requirements:

1.  Microsoft Windows / GNU+Linux / OS X
2.  Java Runtime Environment 1.5 or latter.
Download &#64; [http://java.com](http://java.com)
</div>
<div class="section" id="installing-the-movie-manager">

# Installing the Movie Manager

<dl class="docutils">
<dt>ZIP File:</dt>
<dd>Simply extract the .zip contents to a new directory of your choice.</dd>
<dt>Windows Installer:</dt>
<dd>Double click the installer file and follow the instructions.</dd>
<dt>OS X DMG file:</dt>
<dd>Install like any other DMG file by double clicking and moving the
application to the Applications directory.</dd>
</dl>
</div>
<div class="section" id="how-to-run">

# How to run:

<dl class="docutils">
<dt>Windows:</dt>
<dd>Double click the _MovieManager.exe_ or _MovieManager.jar_. If using the _MovieManager.jar_, the _jar_ extension must be associated with _javaw.exe_.</dd>
<dt>GNU + Linux:</dt>
<dd>**java –jar MovieManager.jar** or if use the shortcuts if installed with cross platform installer.</dd>
<dt>OS X:</dt>
<dd>Using the _Application Bundle_ from the _DMG_ file just double click like any other program. Using the cross-platform installer or ZIP file, start by double clicking the _MovieManager.jar_.</dd>
</dl>
</div>
<div class="section" id="features">

# Features

*   Quick movie filter with advanced search options.
*   List functionality for categorizing movies.
*   Episode functionality for series
*   Multi-add function (Search directories for files)
*   Obtain AVI / OGM / MPEG and DVD (ifo) media file info
*   Automatic download of movie info and series info from IMDb wih Proxy support.
*   Import Text (titles), CSV and Excel.
*   Export to CSV, Excel and HTML.
*   Customizable additional info fields.
*   Changeable layout (Look and Feels)

This product includes software developed by L2FProd.com  ([http://www.L2FProd.com/](http://www.L2FProd.com/)).

Bug Reports / Feature Requests / Contacts

Please report bugs and feature requests in the forum.

Email: [bro3&#64;users.sourceforge.net](mailto:bro3&#64;users.sourceforge.net)
Home Page: [http://xmm.sourceforge.net](http://xmm.sourceforge.net)

</div>
<div class="section" id="links">

# Links:

*   [HSQLDB](http://www.hsqldb.org) - 100% Java Database
*   [MySQL](http://www.mysql.com/products/connector/j) - Connector/J: Official MySQL Java driver
*   [Jexcel](http://www.andykhan.com/jexcelapi) - Java excel driver
*   [BrowserLauncher2](http://browserlaunch2.sourceforge.net) - Browser launcher java library
*   [Launch4j](http://launch4j.sourceforge.net) - Jar wrapper to create windows native exe's
*   [Log4j](http://logging.apache.org/log4j/docs/index.html) - Java logging library
*   [Dotuseful](http://dotuseful.sourceforge.net) - Java library with Swing Tree classes extension
*   [Tablelayout](https://tablelayout.dev.java.net) - TableLayout is a powerful and free layout manager
*   [HttpClient](http://jakarta.apache.org/commons/httpclient) - HTTP java library
*   [JNA](https://jna.dev.java.net/) - Dynamically access native libraries from Java without JNI.
*   [MediaInfo.dll](http://mediainfo.sourceforge.net/en) - Media file parser
*   [JasperReports](http://jasperforge.org/projects/jasperreports) - Report creation utility
*   [IzPack](http://izpack.org) - IzPack Java Installer
*   [Lobo](http://lobobrowser.org/java-browser.jsp) - Java Web Browser: Pure Java browser
*   [Cobra Toolkit](http://lobobrowser.org/cobra.jsp) - Cobra: Java HTML Renderer &amp; Parser
*   [The Castor Project](http://www.castor.org) - Data binding framework for Java
*   [Balloontip](https://balloontip.dev.java.net) - Balloon tips for Java
*   [Jupidator](https://sourceforge.net/projects/jupidator) - Java library for automatic updating
</div>
<div class="section" id="here-are-some-nice-look-feels-for-java">

# Here are some nice Look &amp; Feels for Java:

*   [LaF list at javootoo.com:](http://javootoo.l2fprod.com)
*   [Skinlf](http://www.L2FProd.com) - Skin Look &amp; Feel
*   [NimROD](http://personales.ya.com/nimrod/index-en.html) - NimROD Look &amp; Feel
*   [Substance](https://substance.dev.java.net) - Substance Look &amp; Feel
*   [JTatto](http://www.jtattoo.net) - JTattoo Look &amp; Feel
*   [Office LaF](http://officelnfs.sourceforge.net) - Office Look &amp; Feel
*   [JGoodies LaF](http://www.jgoodies.com) - JGoodies Look &amp; Feel
*   [Kunstoff laf](http://www.incors.org) - Kunstoff Look &amp; Feel

doc_src/Changelog.rst
.. This document is written in reStructuredText format

========================================
   Changelog - MeD's Movie Manager
========================================


Version 2.9.1.3 (20-March-2010)

- General improvements:

  - Enabled anti-aliasing in Swing GUI providing smoother text.
  - When not showing icons in the movie list, the text size is now automatically increased when increasing the list height.
  - The reports in the report generator were showing incorrect page count.
  
- IMDb Library:
  
  - IMDb parser would hang. (Fixed 10.03.2010)


Version 2.9.1.2 (13-February-2010)

- General improvements:

  - Reports for the report generator have been improved and a couple of extra reports were added. 

- Bug fixes:

  - Personal rating was sometimes shown as -1.0 if it hadn't been set.
  - Some strange behaviour with the lists on startup was fixed.
  - Export and Import would fail when using Substance Look and Feel.
  - Report Generator did not function properly.
  

Version 2.9.1.1 (10-February-2010)

- Bug fixes:

  - Export to CSV and Excel was broken.


Version 2.9.1 (07-February-2010)

- New in this version:

  - Added personal rating. This can also be retrieved from IMDb when supplying username and password.
  - Added new Look & Feels (NimROD_).

- Bug fixes:

  - Default values for the additional info fields were saved for the wrong fields when they were rearranged.

- Notes:

  - Because of bugs with additional info default values, the current values may be wrong when imported from old config file.


Version 2.9.0.1 (01-February-2010)

- Bug fixes:
  
  - Config file from v2.8.7 failed to load correctly.


Version 2.9.0 (28-January-2010)

- New in this version:

  - New update notification and automatic download system.

- General improvements:

  - Added better shortcut handling. Press ALT+k on Windows/GNU+Linux or ⌘+k on OSX to see available shortcuts for a window.
  - Updated the HSQLDB driver which will hopefully fix some stability issues with HSQL. 
  - Added right click popup when editing movie/episode to clear additional info values.
  - IMDb dialog now has a field for changing the search text when searching for IMDb info.
  - Added option to remove quotes ("") from series titles.

  - Improvements to the Multi-Add-function:

    - Shows the progress of the import procedure.
    - It's now possible to define custom folders as root devices in the tree.
    - Hardcoded codec info has been replaced with user defined text.
    - Search nfo/txt files for IMDb URL when adding movies.
    - Improved "use folder instead of file name" option to use parent folder if the current folder is named CD1/CD2 etc.
    - New option to prefix movie titles with "_verify_" when using the option "Select First Hit".
    - Added progress bar if expanding a directory takes more than a second.
    - Changed directory separator from '\\' to system file separator in a few places in add multiple movies function.

- Bug fixes:

  - After editing a movie, the lists it is a member of was not shown in the right click popup in the movie list. (Bug ID:2879774)
  - Database converter was unable to convert MySQL or HSQL database to MS Access database.
  - Fixed an issue with selecting multiple entries in the movie list with the SHIFT key.
  - Application would hang if pressing cancel when choosing player after pressing the play button.
  - IMDb scraper would get the episode number and series wrong in some situations.
  - IMDb scraper did not retrieved cast on series.
  - IMDb scraper did not retrieve rating.
  - IMDb scraper could hang (seemingly) indefinitely in some situations (some rare movies).
  

Version 2.8.7 (13-December-2009)

- Bug fixes:

  - Cast was no longer retrieved due to changes in HTML on IMDb.
  - Notes were not saved when editing in main window.
  

Version 2.8.6 (16-July-2009)

- General improvements:

  - No longer have blue border around covers (with some web browsers) when exporting to full HTML.

- Bug fixes:

  - Advanced search was broken with MS Access database.
  - Does not load custom queries when using relative path.


Version 2.8.5 (13-July-2009)

- General improvements:

  - Made the Add/Edit movie panel resizable for those with a row resolution screen.
  - Allow list names containing white space.
  - Added option to skip entries without IMDb id in the "Update IMDb Info" function.

- Bug fixes:

  - Cast was not retrieved after some changes by IMDb.
  - Update covers using "Update IMDb Info" with option "If empty" using MySQL wouldn't always update the cover in the database.
  - IMDb dialog in "Add Multiple Movies" was hiding the "Abort" button.


Version 2.8.4 (04-June-2009)

- General improvements:

  - Support for MediaInfo_ library on 64bit Windows.
  

Version 2.8.3 (01-June-2009)

- Bug fixes:

  - 'Update IMDb Info' would fail for entries with no IMDb id when using Substance L&F.
  - Library was missing making MediaInfo_ library not working at all.


Version 2.8.2 (31-May-2009)

- New 

  - Finding duplicate entries can now be done using the command "DUPLICATES" in the search filter (Not available for MS Access database). Read more about duplicates_.
  - Added Splash Screen during loading of the movie manager for Java 1.6 and later.

- General improvements:

  - Playing media files with space in the name should work better on GNU/Linux. 
  - 'Add multiple movies' now has the option of grouping files together before searching IMDb by right clicking the list elements in list of files to add. Also, files with similar names are shown with the same color in the list of files to add.
  - Support for using MediaInfo_ library on GNU/Linux.
  - Updated the MediaInfo library with better support for mp4 and mkv (Matroska) files.


MovieManager/source/net/sf/xmm/moviemanager/util/SysUtil.java
 /**
 * @(#)SysUtil.java
 * 
 * Copyright (2003) Bro3
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Boston, MA 02111.
 *
 * Contact: bro3@users.sourceforge.net
 **/

package net.sf.xmm.moviemanager.util;

import java.awt.event.InputEvent;
import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.net.URL;
import java.security.AccessController;
import java.security.PrivilegedAction;

import javax.swing.filechooser.FileSystemView;


import net.sf.xmm.moviemanager.MovieManager;
import net.sf.xmm.moviemanager.gui.DialogMovieManager;
import net.sf.xmm.moviemanager.mediainfodll.MediaInfo;
import net.sf.xmm.moviemanager.util.plugins.MovieManagerConfigHandler;

import org.apache.log4j.Logger;


public class SysUtil {

	static Logger log = Logger.getLogger(SysUtil.class);

	/**
	 * Getting the 'root directory' of the app.
	 **/
	public static String getUserDir() {
		
		String path = ""; //$NON-NLS-1$

		try {
			java.net.URL url = FileUtil.class.getProtectionDomain().getCodeSource().getLocation();
			
			File file = new File(java.net.URLDecoder.decode(url.getPath(), "UTF-8")); //$NON-NLS-1$
		
			// If running in a jar file the parent is the root dir 
			if (file.isFile())
				path = file.getParentFile().getAbsolutePath();
			else
				path = file.getAbsolutePath();

		}
		catch (UnsupportedEncodingException e) {
			path = System.getProperty("user.dir"); //$NON-NLS-1$
		}

		if (!path.endsWith(getDirSeparator()))
			path += getDirSeparator();
		
		return path;
	}

	public static String getUserHome() {
		String userHome = (String) AccessController.doPrivileged(
				new PrivilegedAction<Object>() {
					public Object run() {
						return System.getProperty("user.home");
					}
				}
		);
		return userHome;
	}
	
	  
    public static boolean isRestrictedSandbox() {
    	
    	SecurityManager securityManager = System.getSecurityManager();
    	
    	if (securityManager == null) {
    		return false;
    	}
    	
    	try {
    		securityManager.checkPropertiesAccess();
    	} catch (Exception e) {
    		log.debug("Exception:" + e.getMessage());
    		return true;
    	} 
    	
    	return false;
    }
    
    public static boolean isMacAppBundle() {
    	return SysUtil.isMac() && (getJarLocation().indexOf(".app/Contents/Resources") > -1);
    }

    public static String getJarLocation() {
    	
    	String jar = null;
    	
    	jar = MovieManager.class.getProtectionDomain().getCodeSource().getLocation().getPath();
    	jar = jar.replaceAll("%20", " ");
    	return jar;
    }
    
	
	/**
	 * 
	 * @throws Exception 
	 **/
	public static File getConfigDir() throws Exception {
		
		File dir = null;
		
		try {
			
			/* If running in a mac application bundle, we can't write in the application-directory, so we use the /Library/Application Support */
			if (isMac()) {
				String path = System.getProperty("user.home") + "/Library/Application Support/MovieManager/";
				dir = new File(path);
			}
			else if (isWindowsVista() || isWindows7()) {
				
				String path = System.getenv("APPDATA");
				
				if (path == null)
					path = System.getProperty("user.home") + "/" + "Application Data";
					
				dir = new File(path, "MovieManager");
			}

			// Resetting config to local mode
			if (MovieManager.getConfig().getLocalConfigMode()) {
				log.debug("Using localconfigmode for config");
				dir = null;
			}
						
			if (dir != null) {
								
				if (!dir.exists() && !dir.mkdir()) {
					log.error("Could not create settings folder.");
					throw new Exception("Could not create settings folder:" + dir);
				}
			}
			else
				dir = new File(getUserDir(), "config");
						
		}
		catch (Exception e) {
			log.warn("Exception:" + e.getMessage(), e);
		}

		return dir;
	}
	
	public static boolean canWriteToInstallDir() {
		try {
			return FileUtil.canWriteToDir(new File(getUserDir()));
		} catch (Exception e) {
			log.warn("Exception:" + e.getMessage(), e);
		}
		return false;
	}
	
	
	public static URL getConfigURL() {

		URL url = null;
		
		try {
			int appMode = getAppMode();

			// Applet
			if (appMode == 1)
				url = FileUtil.getFileURL("config/Config_Applet.ini", DialogMovieManager.applet);
			else if (appMode == 2) { // Java Web Start
				MovieManagerConfigHandler configHandler = MovieManager.getConfig().getConfigHandler();

				if (configHandler != null)
					url = configHandler.getConfigURL();

			} else {

				String conf = "Config.ini";

				if (SysUtil.isMac())
					url = new File(SysUtil.getConfigDir(), conf).toURI().toURL();
				else {
					
					// First check if Vista or Win7
					// Changed the default location on Vista from program directory to System.getenv("APPDATA")
					
					if (isWindowsVista() || isWindows7()) {
						File config = new File(getConfigDir(), conf);
						
						if (config.isFile())
							url = config.toURI().toURL();
					}
					
					// Find config in install directory
					if (url == null) {

						File conf1 = FileUtil.getFile("config/" + conf);
						File conf2 = FileUtil.getFile(conf);
												
						if (conf1.isFile()) {
							url = FileUtil.getFileURL("config/" + conf);
						}
						else if (conf2.isFile()) {
							url = FileUtil.getFileURL(conf);
						}
					}
					
					// if config file isn't found, check old install dir
					if (url == null) {
					
					// changed default install directory from "MeD's Movie Manager" to "MeDs-Movie-Manager"
						File userDir = new File(getUserDir());

						if (!userDir.getName().equals("MeD's Movie Manager")) {

							// Check old install dir
							File oldInstallDir = new File(userDir.getParentFile(), "MeD's Movie Manager");
							File tmpConfig = new File(oldInstallDir, "config/Config.ini");

							if (tmpConfig.isFile()) {
								url = tmpConfig.toURI().toURL();
							}
							else if ((tmpConfig = new File(oldInstallDir, "Config.ini")).isFile()) {
								url = tmpConfig.toURI().toURL();
							}						
						}
					}
				}
			}
		} catch (Exception e) {
			log.warn("Exception:" + e.getMessage(), e);
		}
		return url;
	}
	
	
	  // 0 = Normal application, 1 = Applet, 2 = Java Web Start
    public static int getAppMode() {

    	int mode = -1;

    	SecurityManager securityManager = System.getSecurityManager();

    	if (securityManager == null) {
    		mode = 0;
		}
    	else {
    		String securityManagerString = securityManager.getClass().getName();

    		if ("com.sun.javaws.security.JavaWebStartSecurity".equals(securityManagerString))
    			mode = 2;
    		else if ("sun.applet.AppletSecurity".equals(securityManagerString))
    			mode = 1;
    	}
    	return mode;
    }

	
	
	public static String getDriveDisplayName(File path) {

		FileSystemView fsv = new javax.swing.JFileChooser().getFileSystemView();

		if (fsv != null) {

			File tmp = path;

			while (tmp.getParentFile() != null)
				tmp = tmp.getParentFile();

			String displayName = fsv.getSystemDisplayName(tmp);

			if (!displayName.trim().equals(""))
				return displayName;

			return "";
		}

		return null;
	}

 
	public static Object getClass(String className) {
		
		if (className != null) {

			try {
				Class<?> classForName = Class.forName(className);
				Object classInstance = classForName.newInstance();
				log.debug("Successfully loaded LoginHandler");
				return classInstance;

			} catch (ClassNotFoundException e) {
				log.error("ClassNotFoundException. Failed to load class " + className);
			} catch (IllegalAccessException e) {
				log.error("IllegalAccessException. Failed to load class " + className);
			} catch (InstantiationException e) {
				log.error("InstantiationException. Failed to load class " + className);
			}
		}
		return null;
	}
	
    
    /* Adds all the files ending in .jar to the classpath */
    public static void includeJarFilesInClasspath(String path) {
            	
        URL url = FileUtil.getFileURL(path);
                
        if (url.toExternalForm().startsWith("http://")) //$NON-NLS-1$
            return;
                
        try {
        	File dir = new File(url.toURI());
        	
        	if (!dir.isDirectory())
        		return;
        	
            File [] jarList = dir.listFiles();
            
            if (jarList != null) {
                
                String absolutePath = ""; //$NON-NLS-1$
                for (int i = 0; i < jarList.length; i++) {
                    
                    absolutePath = jarList[i].getAbsolutePath();
                     
                    if (absolutePath.endsWith(".jar")) { //$NON-NLS-1$
                    	net.sf.xmm.moviemanager.util.ClassPathHacker.addFile(absolutePath);
                        log.debug(absolutePath+ " added to classpath"); //$NON-NLS-1$
                    }
                }
            }
        }
        catch (Exception e) {
            log.error("Exception:" + e.getMessage()); //$NON-NLS-1$
        }
    }
    
    
    public static void cleaStreams(Process p) {
    	
    	if (p == null)
    		return;
    	
    	/**
		 * Clears the streams to avoid having the subprocess hang
		 * @author Bro
		 */
		class StreamHandler extends Thread {

			InputStream inpStr;
			String strType;

			public StreamHandler(InputStream inpStr, String strType) {
				this.inpStr = inpStr;
				this.strType = strType;
			}

			public void run() {
				try {
					InputStreamReader inpStrd = new InputStreamReader(inpStr);
					BufferedReader buffRd = new BufferedReader(inpStrd);
					String line = null;
					String str = "";
					
					while((line = buffRd.readLine()) != null) {
						str += line;
					}
					
					log.debug(line);  
					buffRd.close();
				} catch (IOException e) {
					log.error("1Exception:" + e.getMessage(), e);  
				}
			}
		}
    	
		StreamHandler input = new StreamHandler(p.getInputStream(), "INPUT");
		StreamHandler err = new StreamHandler(p.getErrorStream(), "ERROR");

		input.start();
		err.start();
    }
	
	public static String getLineSeparator() {
		return System.getProperty("line.separator"); //$NON-NLS-1$
	}


	public static String getDirSeparator() {
		return File.separator;
	}

	public static boolean isCtrlPressed(InputEvent event) {
		return ((event.getModifiersEx() & InputEvent.CTRL_DOWN_MASK) == InputEvent.CTRL_DOWN_MASK);
	}

	public static boolean isShiftPressed(InputEvent event) {
		return ((event.getModifiersEx() & InputEvent.SHIFT_DOWN_MASK) == InputEvent.SHIFT_DOWN_MASK);
	}	

	public static boolean isMac() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		return os != null && os.toLowerCase().startsWith("mac"); //$NON-NLS-1$
	}

	public static boolean isOSX() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		return os != null && os.toLowerCase().startsWith("Mac OS X"); //$NON-NLS-1$
	}

	public static boolean isLinux() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		return os != null && os.toLowerCase().startsWith("linux"); //$NON-NLS-1$
	}

	public static boolean isSolaris() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		return os != null && (os.toLowerCase().startsWith("sunos") || os.toLowerCase().startsWith("solaris")); //$NON-NLS-1$ //$NON-NLS-2$
	}

	public static boolean isWindows() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		return os != null && os.toLowerCase().startsWith("windows"); //$NON-NLS-1$
	}

	public static boolean isWindows98() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		return os != null && os.toLowerCase().startsWith("Windows 98"); //$NON-NLS-1$
	}
	
	public static boolean isWindowsXP() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		String osVersion = System.getProperty("os.version"); //$NON-NLS-1$
			
		return os != null && osVersion != null
			&& os.toLowerCase().indexOf("windows") != -1 &&
			osVersion.equals("5.1"); //$NON-NLS-1$
	}

	/**
	 * Bug in Java (not yet fixed in 1.6.0_13) causing System.getProperty("os.name")
	 * to return "Windows XP" on Windows Vista.
	 * System.getProperty("os.version") returns "6.0" on Windows Vista and "5.1" on XP
	 * @return
	 */
	public static boolean isWindowsVista() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		String osVersion = System.getProperty("os.version"); //$NON-NLS-1$
			
		return os != null && osVersion != null
			&& os.toLowerCase().indexOf("windows") != -1 &&
			osVersion.equals("6.0"); //$NON-NLS-1$
	}

	/**
	 * Bug in Java (not yet fixed in 1.6.0_13) causing System.getProperty("os.name")
	 * to return "Windows Vista" on Windows 7.
	 * System.getProperty("os.version") returns "6.1" on Windows 7 and "6.0" on Vista
	 * @return
	 */
	public static boolean isWindows7() {
		String os = System.getProperty("os.name"); //$NON-NLS-1$
		String osVersion = System.getProperty("os.version"); //$NON-NLS-1$
			
		return os != null && osVersion != null
			&& os.toLowerCase().indexOf("windows") != -1 &&
			osVersion.equals("6.1"); //$NON-NLS-1$
	}
	
	
	public static boolean isAMD64() {
		String arch = System.getProperty("os.arch"); //$NON-NLS-1$
		return arch != null  && arch.equals("amd64");
	}
	
	
	public static void openFileLocationOnWindows(File file) {
		try {
			//Desktop.getDesktop().browse(file.toURI());
			Runtime runtime = Runtime.getRuntime();
			runtime.exec("explorer.exe /select,\"" + file.getAbsolutePath() );
		} catch(IOException ioe) {
			ioe.printStackTrace();
			//JOptionPane.showMessageDialog(null, "Could Not Open File Location: " + file.getAbsolutePath());
		}
	}

	  
    public static boolean isCurrentJRES14() {
    	
    	double javaVersion = Double.parseDouble(System.getProperty("java.version").substring(0, 3));
    		
    	if (javaVersion < 1.5) {
    		//log.error("Version:" + javaVersion + " is not supported. Must be 1.5 or higher.");
    		return true;
    	}
    	return false;
    }
    

    public static String getDefaultPlatformBrowser() {
        String browser = "";
        
        if (SysUtil.isWindows())
            browser = "Default";
        else if (SysUtil.isMac())
            browser = "Safari";
        else
            browser = "Firefox";
        
        return browser;
    }
    
    
    public static String getSystemInfo(String separator) {
    
    	String sep = separator == null ? "\\n" : separator;
    	
    	int freeMemory = (int) Runtime.getRuntime().freeMemory()/1024/1024;
    	int totalMemory = (int) Runtime.getRuntime().totalMemory()/1024/1024;
    	int maxMemory = (int) Runtime.getRuntime().maxMemory()/1024/1024;
    	
    	StringBuffer info = new StringBuffer();
    	info.append("Operating System: ").append(System.getProperty("os.name"));
    	info.append(" version: " + System.getProperty("os.version"));
    	info.append(sep);
    	info.append("Architecture: ").append(System.getProperty("os.arch"));
    	
    	info.append(sep);
    	info.append("Java version: ").append(System.getProperty("java.runtime.version"));
    	info.append(sep);
    	info.append("Vendor:").append(System.getProperty("java.vm.specification.vendor"));
    	info.append(sep);
    	
    	info.append("Free VM memory: ").append(freeMemory).append(" MB, ");
    	info.append(sep);
    	info.append("Total VM memory: ").append(totalMemory + " MB");
    	info.append(sep);
    	info.append("Max VM memory: ").append(maxMemory + " MB");
    	
    	return info.toString();
    }
    
    static boolean mediaInfoLibLoaded = false;
    
    public static String getMediaInfoLibVersion() {
    
    	try {
			loadMediaInfoLib();
			String version = MediaInfo.Option_Static("Info_Version");
			return version.split("-")[1].trim();
    	} catch (Exception e) {
			log.warn("Exception:" + e.getMessage());
		}
    	catch (UnsatisfiedLinkError e) {
    		log.debug("UnsatisfiedLinkError:" + e.getMessage());
    		log.debug("Failed to load MediaInfo library");
    	}
    	return null;
    }
    

	public static void loadMediaInfoLib() throws Exception {
		
		if (mediaInfoLibLoaded)
			return;
		
		// Load library on Windows only
		if (SysUtil.isWindows()) {

			String mediaInfoDll = "lib\\MediaInfo\\x86\\MediaInfo.dll";
			
			if (SysUtil.isAMD64()) {
				log.debug("Using MediaInfo library for amd64");
				mediaInfoDll = "lib\\MediaInfo\\amd64\\MediaInfo.dll";
			}
			else {
				log.debug("Using MediaInfo library for x86");
			}
			
			File mediaInfo = new File((FileUtil.getFile(mediaInfoDll)).getPath());

			if (mediaInfo.exists()) {
				LibPathHacker.addDir(FileUtil.getFile("lib").getAbsolutePath());
				System.load(mediaInfo.getAbsolutePath());
			}
			else {
				String error = "";

				if (!mediaInfo.exists()) {
					error += "MediaInfo.dll";
				}

				error = "Following libraries are missing:" + error;
				throw new Exception(error);
			}
		}
		
		/*
		if (SysUtil.isMac()) {

			try {
				System.out.println("load media info for OSX");
				
				String mediaInfoDll = "/Users/bro/Desktop/mediainfo/MediaInfoLib/libmediainfo.0.0.0.dylib";
							
				File mediaInfo = new File((FileUtil.getFile(mediaInfoDll)).getPath());

				
				
				if (mediaInfo.exists()) {
					System.out.println("Loading:" + mediaInfo.getAbsolutePath());
					//LibPathHacker.addDir("/Users/bro/Desktop/mediainfo/MediaInfoLib/");
					//System.load(mediaInfo.getAbsolutePath());
					
					System.out.println("libpath:" + System.getProperty("java.library.path"));
								
					System.loadLibrary("libmediainfo");
				}
				else {
					System.out.println("File doesnt exist");
				}
			} catch (UnsatisfiedLinkError e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}
		*/
	}
	
    
    /**
     * Checks if the current running JRE version is at least 1.6
     * @return   true if JRE is at least 1.6
     */
    public static boolean isAtLeastJRE6() {

    	double javaVersion = Double.parseDouble(System.getProperty("java.version").substring(0, 3));

    	if (javaVersion >= 1.6) {
    		return true;
    	}
    	return false;
    }

	
	public static String colourToString(java.awt.Color c) {
	    return Integer.toHexString(0xFF000000 | c.getRGB()).substring(2);
	}
 
}
